x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/Match.java

```diff
@@ -65,29 +72,39 @@ protected TypeResolution resolveType() {
             return new TypeResolution("Unresolved children");
         }
 
-        TypeResolution resolution = isStringAndExact(field, sourceText(), ParamOrdinal.FIRST);
-        if (resolution.unresolved()) {
-            return resolution;
-        }
-
-        int index = 1;
-        for (Expression regex : patterns) {
+        TypeResolution resolution;
+        int index = 0;
+        List<Integer> nonFoldables = new ArrayList<>(2);
+        for (Expression e : children()) {
             // Currently we have limited enum for ordinal numbers
             // So just using default here for error messaging
-            resolution = isStringAndExact(regex, sourceText(), ParamOrdinal.fromIndex(index));
+            ParamOrdinal paramOrd = ParamOrdinal.fromIndex(index);
+            resolution = isStringAndExact(e, sourceText(), paramOrd);
             if (resolution.unresolved()) {
                 return resolution;
             }
-
-            resolution = isFoldable(regex, sourceText(), ParamOrdinal.fromIndex(index));
+            resolution = isFoldable(e, sourceText(), paramOrd);
             if (resolution.unresolved()) {
-                break;
+                nonFoldables.add(index);
             }
-
             index++;
         }
 
-        return resolution;
+        if (nonFoldables.size() == 2) {
+            StringBuilder sb = new StringBuilder(format(null, "only one argument of [{}] must be non-constant but multiple found: ",
```

Comment:
I would emphasize the two "versions" of `match` function, rather than mentioning that one argument needs to be a field.
Something more aligned with the (to be updated) documentation.

For example "either first or last argument of `match` must be a non-constant" (or around the same idea).

Reply:
But now we allow the field to be anywhere, should we change and be stricter and allow only first and last?

Reply:
Based on the [docs](https://eql.readthedocs.io/en/latest/query-guide/functions.html#match) my understand is the field (source) would have to be first and the rest would be patterns.
There's no mention of them being literals but the code does seem to indicate that.

@aleksmaus @rw-access can you please advise?

Reply:
Yes, the rest must be literal strings. I believe we currently accept expressions that fold to strings as well.

I can update the documentation in a few places to reflect that. The rationale is that building regular expressions on the fly generally lead to poor performance or malformed expressions. For some of the existing EQL implementations, we need to know at compile time, so that the state machine can be built only once.

Reply:
The same restrictions apply to `match`, `cidrMatch` and `wildcard`.
At one point in the past, the function signature used to look like this: `match(<regex>, <expr>)`, but that's been abandoned and now it's a variadic function of the form `match(<expr>, <regex1>, <regex2>, ...)`. There may have been some tests that were of the old form that didn't get changed because they were being supported for backwards compatibility.  But I think it's okay to drop the ambiguous representation and stick with `match(<expr>, <regex1>, <regex2>, ...)`